src/panels/lovelace/common/directives/action-handler-directive.ts

```diff
@@ -0,0 +1,216 @@
+  actionHandler?: boolean;
+}
+
+declare global {
+  interface HASSDomEvents {
+    action: ActionHandlerEvent;
+  }
+}
+
+class ActionHandler extends HTMLElement implements ActionHandler {
+  public holdTime: number;
+  public ripple: any;
+  protected timer: number | undefined;
+  protected held: boolean;
+  protected cooldownStart: boolean;
+  protected cooldownEnd: boolean;
+  private dblClickTimeout: number | undefined;
+
+  constructor() {
+    super();
+    this.holdTime = 500;
+    this.ripple = document.createElement("mwc-ripple");
+    this.timer = undefined;
+    this.held = false;
+    this.cooldownStart = false;
+    this.cooldownEnd = false;
+  }
+
+  public connectedCallback() {
+    Object.assign(this.style, {
+      position: "absolute",
+      width: isTouch ? "100px" : "50px",
+      height: isTouch ? "100px" : "50px",
+      transform: "translate(-50%, -50%)",
+      pointerEvents: "none",
+    });
+
+    this.appendChild(this.ripple);
+    this.ripple.primary = true;
+
+    [
+      "touchcancel",
+      "mouseout",
+      "mouseup",
+      "touchmove",
+      "mousewheel",
+      "wheel",
+      "scroll",
+    ].forEach((ev) => {
+      document.addEventListener(
+        ev,
+        () => {
+          clearTimeout(this.timer);
+          this.stopAnimation();
+          this.timer = undefined;
+        },
+        { passive: true }
+      );
+    });
+  }
+
+  public bind(element: ActionHandlerElement, options) {
+    if (element.actionHandler) {
+      return;
+    }
+    element.actionHandler = true;
+
+    element.addEventListener("contextmenu", (ev: Event) => {
+      const e = ev || window.event;
+      if (e.preventDefault) {
+        e.preventDefault();
+      }
+      if (e.stopPropagation) {
+        e.stopPropagation();
+      }
+      e.cancelBubble = true;
+      e.returnValue = false;
+      return false;
+    });
+
+    const clickStart = (ev: Event) => {
+      if (this.cooldownStart) {
+        return;
+      }
+      this.held = false;
+      let x;
+      let y;
+      if ((ev as TouchEvent).touches) {
+        x = (ev as TouchEvent).touches[0].pageX;
+        y = (ev as TouchEvent).touches[0].pageY;
+      } else {
+        x = (ev as MouseEvent).pageX;
+        y = (ev as MouseEvent).pageY;
+      }
+
+      if (options.hasHold) {
+        this.timer = window.setTimeout(() => {
+          this.startAnimation(x, y);
+          this.held = true;
+        }, this.holdTime);
+      }
+
+      this.cooldownStart = true;
+      window.setTimeout(() => (this.cooldownStart = false), 100);
+    };
+
+    const clickEnd = (ev: Event) => {
+      if (
+        this.cooldownEnd ||
+        (["touchend", "touchcancel"].includes(ev.type) &&
+          this.timer === undefined)
+      ) {
+        return;
+      }
+      clearTimeout(this.timer);
+      this.stopAnimation();
+      this.timer = undefined;
+      if (this.held) {
+        fireEvent(this, "action", { action: "hold" });
```

Comment:
The event should be fired from the `element`, not `this`.
```suggestion
        fireEvent(element, "action", { action: "hold" });
```

Reply:
there it is; knew I missed something simple